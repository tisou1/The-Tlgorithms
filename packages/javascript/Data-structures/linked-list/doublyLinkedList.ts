import type { DoubleNode } from './types'

class Node<T> {
  data: T
  next: DoubleNode<T> | null
  prev: DoubleNode<T> | null
  constructor(element: T) {
    this.data = element
    this.next = this.prev = null
  }
}

class DoublyLinkedList<T> {
  length: number
  head: DoubleNode<T> | null
  tail: DoubleNode<T> | null
  constructor() {
    this.length = 0
    this.head = null
    this.tail = null
  }

  size = () => this.length
  
  //添加元素
  append(data: T) {
    const node = new Node(data)

    if(!this.head) {
      this.head = this.tail = node
    } else {
      node.prev = this.tail
      this.tail!.next = node
      this.tail = node
    }

    this.length++

    return this.size()
  }

  //获取元素
  getElementAt(index: number) {
    if(index >=0 && index <= this.length) {
      //正常范围值
      let currentNode = this.head
      let currentIndex = 0
      while(currentIndex < index) {
        currentNode = currentNode!.next
        currentIndex++
      }

      return currentNode
    } else {
      return null
    }
  }

  insert(index: number, data: T) {
    const node = new Node(data)
    //获取目标元素前一个
    const previousNode = this.getElementAt(index - 1)


    node.next = previousNode!.next
    previousNode!.next!.prev = node as DoubleNode<T>

    previousNode!.next = node
    node.prev = previousNode

    return ++this.length

  }

  // Convert list to array
  toArray () {
    const arr: T[] = []
    let current = this.head

    while (current) {
      arr.push(current.data)
      current = current.next
    }

    return arr
  }
}

export { DoublyLinkedList }