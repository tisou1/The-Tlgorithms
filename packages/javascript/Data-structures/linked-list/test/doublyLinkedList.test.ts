import { describe, it, expect, beforeEach } from 'vitest'
import { DoublyLinkedList } from '../doublyLinkedList'

describe('check DoublyLinkedList', () => {
  let list
  beforeEach(() => {
    list = new DoublyLinkedList<unknown>()
  })

  it('check size', () => {
    expect(list.size()).toBe(0)
  })

  it('check append', () => {
    expect(list.append(11)).toBe(1)
    expect(list.append(22)).toBe(2)
    expect(list.append(33)).toBe(3)
    expect(list.toArray()).toEqual([11,22,33])
  })

  it('check getElementAt', () => {
    expect(list.append(11)).toBe(1)
    expect(list.append(22)).toBe(2)
    expect(list.append(33)).toBe(3)

    expect(list.getElementAt(1)).toMatchSnapshot()
  })

  it('check insert', () => {
    expect(list.append(10)).toBe(1)
    expect(list.append(20)).toBe(2)
    expect(list.append(30)).toBe(3)
    expect(list.insert(1,333)).toEqual(4)
    expect(list.insert(3, 444)).toEqual(5)
    expect(list.toArray()).toEqual([10, 333, 20, 444, 30])
    expect(list.toArray()).toMatchInlineSnapshot(`
      [
        10,
        333,
        20,
        444,
        30,
      ]
    `)
  })
})