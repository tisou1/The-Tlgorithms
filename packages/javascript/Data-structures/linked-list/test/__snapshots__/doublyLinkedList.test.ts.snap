// Vitest Snapshot v1

exports[`check DoublyLinkedList > check getElementAt 1`] = `
Node {
  "data": 22,
  "next": Node {
    "data": 33,
    "next": null,
    "prev": [Circular],
  },
  "prev": Node {
    "data": 11,
    "next": [Circular],
    "prev": null,
  },
}
`;
